### DISCUSSION TITLE
Redundant "value_error.missing" messages for body params in request with empty body.

### DISCUSSION THREAD

**Discussion by dabula-s** [fastapi/fastapi#8711] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Body, Query


app = FastAPI()


@app.post('/test-body-params-in-post')
def test_endpoint1(param1: str = Body(...),
                  param2: str = Body(None),
                  param3: str = Body('bla')):
    return [param1, param2, param3]


@app.post('/test-query-params-in-post')
def test_endpoint2(param1: str = Query(...),
                  param2: str = Query(None),
                  param3: str = Query('bla')):
    return [param1, param2, param3]


@app.post('/test-query-and-body-params-in-post')
def test_endpoint3(param1: str = Query(...),
                  param2: str = Query(None),
                  param3: str = Query('bla'),
                  param4: str = Body(...),
                  param5: str = Body(None),
                  param6: str = Body('bla')):
    return [param1, param2, param3, param4, param5, param6]
```

### Description

So, lets start making empty requests via requests and see what fastapi returns in error message.
```
requests.post('http://localhost/test-body-params-in-post').text
'{"detail":[{"loc":["body","param1"],"msg":"field required","type":"value_error.missing"}]}'
```
Hm, nothing strange. Ok. Continue. What does happen if we send an empty string instead of empty json (aka {})?
```
requests.post('http://localhost/test-body-params-in-post', json="").text
'{"detail":[{"loc":["body","param1"],"msg":"field required","type":"value_error.missing"},{"loc":["body","param2"],"msg":"field required","type":"value_error.missing"},{"loc":["body","param3"],"msg":"field required","type":"value_error.missing"}]}'
```
Looks weird. We have only one required parameter but fastapi returns "field required" for all.

```
requests.post('http://localhost/test-query-and-body-params-in-post', json={}).text
'{"detail":[{"loc":["query","param1"],"msg":"field required","type":"value_error.missing"},{"loc":["body","param4"],"msg":"field required","type":"value_error.missing"}]}'

requests.post('http://localhost/test-query-and-body-params-in-post', json="").text
'{"detail":[{"loc":["query","param1"],"msg":"field required","type":"value_error.missing"},{"loc":["body","param4"],"msg":"field required","type":"value_error.missing"},{"loc":["body","param5"],"msg":"field required","type":"value_error.missing"},{"loc":["body","param6"],"msg":"field required","type":"value_error.missing"}]}'

```

### Operating System

Linux

### FastAPI Version

0.65.3, 0.70.1

### Python Version

Python 3.9.9

**Comment by dabula-s** [comment]
this behavior exists also in cases with json like `"[]", "[{}]", etc.`

**Comment by insomnes** [comment]
>Hm, nothing strange. Ok. Continue. What does happen if we send an empty string instead of empty json (aka {})?

There problem is in how you send your data, when you write:
```
requests.post('http://localhost/test-body-params-in-post', json="").text
```
In JSON you will actually get:
```
json.dumps("")
'""'
```
Cause you tell json to encode empty, but still featured python string object
But if you send
```
requests.post('http://localhost/test-body-params-in-post', data="".encode("utf-8")).text
```
Result will be normal:
```
{"detail":[{"loc":["body","param1"],"msg":"field required","type":"value_error.missing"}]}
```

Compare:
```
json.dumps("").encode("utf-8")
=>
b'""'
"".encode("utf-8")
=>
b''
```

**Comment by dabula-s** [comment]
> > Hm, nothing strange. Ok. Continue. What does happen if we send an empty string instead of empty json (aka {})?
> 
> There problem is in how you send your data, when you write:
> 
> ```
> requests.post('http://localhost/test-body-params-in-post', json="").text
> ```
> 
> In JSON you will actually get:
> 
> ```
> json.dumps("")
> '""'
> ```
> 
> Cause you tell json to encode empty, but still featured python string object But if you send
> 
> ```
> requests.post('http://localhost/test-body-params-in-post', data="".encode("utf-8")).text
> ```
> 
> Result will be normal:
> 
> ```
> {"detail":[{"loc":["body","param1"],"msg":"field required","type":"value_error.missing"}]}
> ```
> 
> Compare:
> 
> ```
> json.dumps("").encode("utf-8")
> =>
> b'""'
> "".encode("utf-8")
> =>
> b''
> ```

yep, thats because json.dumps(""). and its not my problem. i guess i could send everything and should get logically expected response.

**Comment by insomnes** [comment]
I don't think so, when you send invalid JSON object type you will get corresponding error message

**Comment by dabula-s** [comment]
> I don't think so, when you send invalid JSON object type you will get corresponding error message

and it would be logically expected response.

**Comment by dabula-s** [comment]
> this behavior exists also in cases with json like `"[]", "[{}]", etc.`

@insomnes you can try to send these through the api docs and see the same response where all non required fields became required.

**Comment by dabula-s** [comment]
![image](https://user-images.githubusercontent.com/22279718/147075785-64ed04c5-b09e-4553-9298-a1478aaa0944.png)

**Comment by insomnes** [comment]
I've checked it on pydantic models and it's behavior is different, right as you expect.
Need to check sources to see what causes difference.

**Comment by insomnes** [comment]
Interesting, if body decoded as json, and object type in it not support `received_body.get(field.alias)` method, than even not required body parameters will be added as missing

**Comment by Kernel5555** [comment]
`@router.post(f'{settings.API_URL_PREFIX}/user/create', response_model = User_Pydantic)

async def create_user(req: UserIn_Pydantic):

    user_obj = await User.create(
        email = req.email,
        phone = req.phone,
        password = bcrypt.hash(req.password),
        first_name = req.first_name,
        last_name = req.last_name
    )

    await user_obj.save()

    return await User_Pydantic.from_tortoise_orm(user_obj)`

@SeriyVol4ishe, I had this issues ... was actually the cause of failing to await for a returning object after save.

**Comment by YuriiMotov** [comment]
Still reproducible in FastAPI 0.115.13.

Simplified test case:
```python
from fastapi import Body, FastAPI
from fastapi.testclient import TestClient

app = FastAPI()

@app.post("/test")
def ep(
    param1: str = Body(...),  # Only this parameter is required
    param2: str = Body(None),
    param3: str = Body("bla"),
):
    return "ok"

def test_():
    client = TestClient(app)
    response = client.post("/test", json={})
    assert response.status_code == 422
    respose_json = response.json()

    assert len(respose_json["detail"]) == 1, respose_json  # AssertionError: assert 3 == 1

    assert respose_json["detail"][0]["type"] == "missing"
    assert respose_json["detail"][0]["loc"] == ["body", "param1"]
```

Only `param` is required, but if body is empty, than error response includes  `'type': 'missing', 'msg': 'Field required',` for all fields:
```
[
    {
        "type": "missing",
        "loc": ["body", "param1"],
        "msg": "Field required",
        "input": None,
    },
    {
        "type": "missing",
        "loc": ["body", "param2"],
        "msg": "Field required",
        "input": None,
    },
    {
        "type": "missing",
        "loc": ["body", "param3"],
        "msg": "Field required",
        "input": None,
    },
]
```

At the same time, passing `json={}` works as expected
